{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ef4e41c5-e40d-471e-b3eb-1bfcaae6c0c2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import pickle\n",
    "import random\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "sys.path.append('../src/')\n",
    "from configs import *\n",
    "from utils_optimization import *\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "from matplotlib import rcParams\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "rcParams.update(fig_params)\n",
    "import seaborn as sns\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c732a3d2-ce2e-437f-a11d-5d57291672fd",
   "metadata": {},
   "source": [
    "## Load data and models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2623b787-13a3-45d4-a742-0825ee2e4e07",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Featurizer\n",
    "featurizer = Featurizer(data_folder=os.path.join('..', 'data'))\n",
    "\n",
    "## Model\n",
    "with open(os.path.join('..', 'models', 'efficient_model.pickle'), 'rb') as handle:\n",
    "    models = pickle.load(handle)\n",
    "best_iteration = models['best_iteration']\n",
    "n_models = 1 #len(models['models_ensemble'].boosters)\n",
    "\n",
    "## Train dataset\n",
    "train, features = load_dataset()\n",
    "\n",
    "## Selected features\n",
    "selected_features = ['CTDC_charge.G1',\n",
    "                     'CTDD_hydrophobicity_ARGP820101.1.residue50',\n",
    "                     'SOCNumber_gGrantham.lag1',\n",
    "                     'DDE_LA',\n",
    "                     'Geary_ANDN920101.lag2',\n",
    "                     'DDE_SP',\n",
    "                     'NetC',\n",
    "                     'ASDC_VG',\n",
    "                     'CTDC_solventaccess.G3',\n",
    "                     'APAAC_Pc1.V']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e81630a4-45fa-4383-a702-6c71d77f7aa7",
   "metadata": {},
   "source": [
    "## Optimization loop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3c77ba99-8dea-42e6-a339-4eb83cdd7372",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Penetrability of target: 0.9366376658161918\n",
      "Anomaly of target: [0.99796328]\n",
      "Generation: 1 -- Best solution: RNSYRKSPSRRVH -- Best fitness: 0.20588 -- Prob_penetration: 0.98 -- Anomaly score: 1.04 -- Different letters from target: 2\n",
      "Generation: 2 -- Best solution: RNSYQKPPQRRNR -- Best fitness: 0.19118 -- Prob_penetration: 0.88 -- Anomaly score: 0.99 -- Different letters from target: 3\n",
      "Generation: 3 -- Best solution: RNSYRKSPSRRYR -- Best fitness: 0.11765 -- Prob_penetration: 0.96 -- Anomaly score: 0.99 -- Different letters from target: 1\n",
      "Generation: 3 -- Best solution: RNSYRKSPSRRYR -- Best fitness: 0.11764705882352944 -- Different letters from target: 1\n",
      "Best solution with 1 different letters from target: RNSYRKSPSRRYR -- Prob_penetration: 0.96 -- Anomaly score: 0.99\n",
      "Best solution with 2 different letters from target: RDSYRKSPSARNR -- Prob_penetration: 0.92 -- Anomaly score: 1.26\n",
      "Best solution with 3 different letters from target: RNSYQKPPQRRNR -- Prob_penetration: 0.88 -- Anomaly score: 0.99\n",
      "Best solution with 4 different letters from target: RTSYKKSPRQRNR -- Prob_penetration: 0.94 -- Anomaly score: 1.09\n",
      "Best solution with 5 different letters from target: QNKYRKSPKHRND -- Prob_penetration: 0.99 -- Anomaly score: 1.07\n",
      "Best solution with 6 different letters from target: RNSFRKSSKSRGK -- Prob_penetration: 0.82 -- Anomaly score: 0.98\n",
      "Best solution with 7 different letters from target: VMKQRKHPPRRDR -- Prob_penetration: 1.00 -- Anomaly score: 0.99\n",
      "Best solution with 8 different letters from target: RNKYCRNWGKRPR -- Prob_penetration: 0.85 -- Anomaly score: 1.08\n",
      "Best solution with 9 different letters from target: EGWYRKRMKCQAR -- Prob_penetration: 0.69 -- Anomaly score: 1.04\n",
      "Best solution with 10 different letters from target: RSTNRSNWTRILT -- Prob_penetration: 0.04 -- Anomaly score: 1.02\n"
     ]
    }
   ],
   "source": [
    "TARGET = 'RNSYRKSPSRRNR'\n",
    "len_gnome = len(TARGET)\n",
    "max_iter = 50\n",
    "size_population = 100\n",
    "genes = '''ARNDCEQGHILKMFPSTWYV'''\n",
    "\n",
    "# Training features for anomaly detection\n",
    "feat_train = train[selected_features]\n",
    "clf_anomaly = LocalOutlierFactor(n_neighbors=20, novelty=True, contamination=0.1)\n",
    "clf_anomaly.fit(feat_train)\n",
    "\n",
    "# Sequence alignment settings\n",
    "aligner = Align.PairwiseAligner()\n",
    "aligner.substitution_matrix = substitution_matrices.load(\"BLOSUM62\")\n",
    "aligner.target_internal_open_gap_score = -20\n",
    "aligner.query_internal_open_gap_score = -20\n",
    "\n",
    "generation = 1\n",
    "population = []\n",
    "\n",
    "optimization_info = {'Pop_size': size_population, 'max_iter': max_iter}\n",
    "\n",
    "params_fit = {'obj': my_fitness, 'target_ligand': TARGET, 'aligner': aligner, 'model': models, 'clf_anomaly': clf_anomaly, \n",
    "             'featurizer': featurizer, 'best_iteration': best_iteration, 'n_models': n_models}\n",
    "\n",
    "features_TARGET = featurizer.compute_features(TARGET)\n",
    "print(f\"Penetrability of target: {pred_penetration(features_TARGET, models['models_refit'], best_iteration, n_models)}\")\n",
    "print(f\"Anomaly of target: {anomaly_score(features_TARGET, clf_anomaly)}\")\n",
    "\n",
    "# Create initial population\n",
    "max_diff_pct = 50  # set the percentage of different letters, user can change this value\n",
    "for _ in range(size_population):\n",
    "    gnome = Individual.create_gnome(len_gnome, genes, TARGET, max_diff_pct)\n",
    "    population.append(Individual(gnome, params_fit))\n",
    "\n",
    "# Define stopping criteria thresholds\n",
    "penetration_threshold = 0.95\n",
    "fitness_threshold = 0.1\n",
    "anomaly_threshold = 1.3\n",
    "\n",
    "best_sequences = {}\n",
    "\n",
    "while generation <= max_iter:\n",
    "    # Sort the population in increasing order of fitness score\n",
    "    population = sorted(population, key=lambda x: x.fitness)\n",
    "\n",
    "    for individual in population:\n",
    "        if individual.num_diff_residues not in best_sequences:\n",
    "            best_sequences[individual.num_diff_residues] = individual\n",
    "        else:\n",
    "            if individual.fitness < best_sequences[individual.num_diff_residues].fitness:\n",
    "                best_sequences[individual.num_diff_residues] = individual\n",
    "\n",
    "    # If the best solution in the population meets the stopping criteria, break the loop\n",
    "    best_solution = population[0]\n",
    "    features_best = featurizer.compute_features(''.join(best_solution.chromosome))\n",
    "    penetration_score = pred_penetration(features_best, models['models_refit'], best_iteration, n_models)\n",
    "    anomaly_score_value = anomaly_score(features_best, clf_anomaly)[0]\n",
    "\n",
    "    # Including the condition for num_diff_residues\n",
    "    if (penetration_score > penetration_threshold and anomaly_score_value < anomaly_threshold and best_solution.num_diff_residues <= 1):\n",
    "        break\n",
    "    # Including the condition for num_diff_residues\n",
    "    if (best_solution.num_diff_residues == 0):\n",
    "        break\n",
    "    # Generate new offsprings for new generation\n",
    "    new_generation = []\n",
    "\n",
    "    # Perform Elitism, that mean 10% of fittest population\n",
    "    # goes to the next generation\n",
    "    s = int((10*size_population) / 100)\n",
    "    new_generation.extend(population[:s])\n",
    "\n",
    "    # From 50% of fittest population, Individuals\n",
    "    # will mate to produce offspring\n",
    "    s = int((90*size_population) / 100)\n",
    "    for _ in range(s):\n",
    "        parent1 = random.choice(population[:50])\n",
    "        parent2 = random.choice(population[:50])\n",
    "        child = parent1.mate(parent2, genes)\n",
    "        new_generation.append(child)\n",
    "\n",
    "    population = new_generation\n",
    "\n",
    "    print(f\"Generation: {generation} -- \"\n",
    "          f\"Best solution: {''.join(population[0].chromosome)} -- \"\n",
    "          f\"Best fitness: {population[0].fitness:.5f} -- \"\n",
    "          f\"Prob_penetration: {penetration_score:.2f} -- \"\n",
    "          f\"Anomaly score: {anomaly_score_value:.2f} -- \"\n",
    "          f\"Different letters from target: {population[0].num_diff_residues}\"\n",
    "          )\n",
    "\n",
    "    generation += 1\n",
    "\n",
    "print(f\"Generation: {generation} -- \"\n",
    "      f\"Best solution: {''.join(population[0].chromosome)} -- \"\n",
    "      f\"Best fitness: {population[0].fitness:.5f} -- \"\n",
    "      f\"Prob_penetration: {penetration_score:.2f} -- \"\n",
    "      f\"Anomaly score: {anomaly_score_value:.2f} -- \"\n",
    "      f\"Different letters from target: {population[0].num_diff_residues}\"\n",
    "      )\n",
    "\n",
    "features_opt = featurizer.compute_features(''.join(population[0].chromosome))\n",
    "distance = compute_distance(aligner, TARGET, ''.join(population[0].chromosome))\n",
    "\n",
    "print(f\"Generation: {generation} -- \"\n",
    "      f\"Best solution: {''.join(population[0].chromosome)} -- \"\n",
    "      f\"Best fitness: {population[0].fitness} -- \"\n",
    "      f\"Different letters from target: {population[0].num_diff_residues}\"\n",
    "      )\n",
    "\n",
    "# Print the best sequence for each number of different letters\n",
    "for num_diff_residues, individual in sorted(best_sequences.items()):\n",
    "    print(f\"Best solution with {num_diff_residues} different letters from target: \"\n",
    "          f\"{''.join(individual.chromosome)} -- \"\n",
    "          f\"Prob_penetration: {pred_penetration(featurizer.compute_features(''.join(individual.chromosome)), models['models_refit'], best_iteration, n_models):.2f} -- \"\n",
    "          f\"Anomaly score: {anomaly_score(featurizer.compute_features(''.join(individual.chromosome)), clf_anomaly)[0]:.2f}\"\n",
    "          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1b558a2-4a56-4103-bb4c-f64c661aa63a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
